systemd/nslogin/nslogin.c

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 a
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// That's GCC's way of letting us access GNU extensions. We either define this macro in this source file, before all
+// includes or on the Makefile.
+// NOLINTNEXTLINE(bugprone-reserved-identifier)
+#define _GNU_SOURCE
+
+#include <fcntl.h>
+#include <inttypes.h>
+#include <linux/limits.h>
+#include <pwd.h>
+#include <sched.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <systemd/sd-bus.h>
+#include <unistd.h>
+
+/// Returns systemd real PID.
+pid_t find_systemd(void);
+
+/// Returns true when the process PID is ready.
+bool wait_on_systemd(void);
+
+/// Enters in the mount and pid namespaces of the process PID, while preserving PWD.
+/// Returns 0 on success.
+int enter_target_ns(int PID);
+
+/// Stops this process in favor or running it as child.
+void continue_as_child(void);
+
+#define RESTART_MSG "Please terminate this instance by running \"wsl -t <distro>\" from Windows shell and try again.\n"
+
+int main(int argc, char *argv[]) {
+    uid_t uid = getuid();
+    gid_t gid = getgid();
+#ifndef NDEBUG
+    printf("Starting nslogin with arguments: ");
+    for (int i = 0; i < argc; i++) {
+        printf("%s ", argv[i]);
+    }
+    putc('\n', stdout);
+    // print euid
+    printf("euid: %d\n", geteuid());
+    printf("uid: %d\n", getuid());
+#endif
+    // Wait for systemd to be ready
+    pid_t systemdPid = find_systemd();
+    if (systemdPid == 0) {
+        fprintf(stderr, "Systemd is not running. " RESTART_MSG);
+        exit(1);
+    }
+    if (!wait_on_systemd()) {
+        fprintf(stderr, RESTART_MSG);
```

Comment:
Prefix with an ERROR: or something obviously it’s failing.
Also, I would add something like ("systemd is running, but not able to proceed boot") or along those lines